src/symbols.c

```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+
+	/* TreeView and its model */
+	store = gtk_list_store_new(N_COLUMNS, GDK_TYPE_PIXBUF, G_TYPE_STRING, TM_TYPE_TAG);
+	tag_goto_tree_view = gtk_tree_view_new_with_model(GTK_TREE_MODEL(store));
+	gtk_widget_set_can_focus(tag_goto_tree_view, TRUE);
+	gtk_container_add(GTK_CONTAINER(scroller), tag_goto_tree_view);
+	g_signal_connect(tag_goto_tree_view, "focus-out-event", G_CALLBACK(on_focus_out), tag_goto_popup);
+	g_signal_connect(tag_goto_tree_view, "row-activated", G_CALLBACK(on_row_activated), NULL);
+	g_signal_connect(tag_goto_tree_view, "key-press-event", G_CALLBACK(on_key_pressed), NULL);
+
+	selection = gtk_tree_view_get_selection(GTK_TREE_VIEW(tag_goto_tree_view));
+	gtk_tree_selection_set_mode(selection, GTK_SELECTION_SINGLE);
+	gtk_tree_view_set_headers_visible(GTK_TREE_VIEW(tag_goto_tree_view), FALSE);
+	gtk_tree_view_set_reorderable(GTK_TREE_VIEW(tag_goto_tree_view), FALSE);
+
+	/* Columns */
+	column = gtk_tree_view_column_new();
+	gtk_tree_view_column_set_sizing(column, GTK_TREE_VIEW_COLUMN_FIXED);
+
+	renderer = gtk_cell_renderer_pixbuf_new();
+	gtk_tree_view_column_pack_start(column, renderer, FALSE);
+	gtk_tree_view_column_add_attribute(column, renderer, "pixbuf", PIXBUF_COLUMN);
+
+	renderer = gtk_cell_renderer_text_new();
+	gtk_tree_view_column_pack_start(column, renderer, TRUE);
+	gtk_tree_view_column_add_attribute(column, renderer, "markup", TEXT_COLUMN);
+
+	gtk_tree_view_append_column(GTK_TREE_VIEW(tag_goto_tree_view), column);
+
+	gtk_widget_show_all(frame);
+}
+
+
+static void show_goto_popup(GPtrArray *tags, gboolean have_best)
+{
+	gboolean first = TRUE;
+	GtkTreePath *path;
+	GtkTreeModel *model;
+	GtkTreeIter iter;
+	TMTag *tmtag;
+	gint i;
+
+	if (!tag_goto_popup)
+		create_goto_popup();
+
+	model = gtk_tree_view_get_model(GTK_TREE_VIEW(tag_goto_tree_view));
+	gtk_list_store_clear(GTK_LIST_STORE(model));
+
+	foreach_ptr_array(tmtag, i, tags)
+	{
+		gchar *fname = g_path_get_basename(tmtag->file->file_name);
+		gchar *text;
+
+		if (first && have_best)
+			text = g_strdup_printf("<b>%s: %lu</b>", fname, tmtag->line);
+		else
+			text = g_strdup_printf("%s: %lu", fname, tmtag->line);
+
+		gtk_list_store_insert_with_values(GTK_LIST_STORE(model), &iter, -1,
+				PIXBUF_COLUMN, get_tag_icon(get_tag_class(tmtag)),
+				TEXT_COLUMN, text,
+				TAG_COLUMN, tmtag, -1);
+
+		g_free(text);
+		g_free(fname);
+		first = FALSE;
+	}
+
+	path = gtk_tree_path_new_first();
+	gtk_tree_view_set_cursor (GTK_TREE_VIEW(tag_goto_tree_view), path, NULL, FALSE);
+	gtk_tree_path_free(path);
+
+	gtk_window_present(GTK_WINDOW(tag_goto_popup));
+}
+
+
+static gint compare_tags(gconstpointer ptr1, gconstpointer ptr2)
+{
+	gint res;
+	TMTag *t1 = *((TMTag **) ptr1);
+	TMTag *t2 = *((TMTag **) ptr2);
+
+	res = g_strcmp0(t1->file->short_name, t2->file->short_name);
+	if (res != 0)
+		return res;
+	return t1->line - t2->line;
+}
+
+
+static TMTag *find_best_goto_tag(GPtrArray *tags)
+{
+	GeanyDocument *doc = document_get_current();
```

Comment:
I'd rather pass this as an argument, lowering the environmental factor of this function